### ISSUE TITLE
concatenated_forward when self.ref_model is not provided

### ISSUE THREAD

**Issue by mjbooo** [huggingface/trl#1500]
https://github.com/huggingface/trl/blob/0ee349dcd43b0f4b3169449f16751c38ac4a609f/trl/trainer/dpo_trainer.py#L758C1-L771C76

```
 def compute_reference_log_probs(self, padded_batch: Dict) -> Dict:
        """Computes log probabilities of the reference model for a single padded batch of a DPO specific dataset."""
        compte_ref_context_manager = torch.cuda.amp.autocast if self._peft_has_been_casted_to_bf16 else nullcontext

        # compute reference logps
        with torch.no_grad(), compte_ref_context_manager():
            if self.ref_model is None:
                with self.null_ref_context():
                    (
                        reference_chosen_logps,
                        reference_rejected_logps,
                        _,
                        _,
                    ) = self.concatenated_forward(self.model, padded_batch)
```

The last line of the code, concatenated_forward(self.model, ...), appears to get log probabilities for chosen and rejected pairs using the **updated model (=self.model)**. 
However, this process should ideally be conducted with a model that has not been updated at all.

Am I understanding this correctly? If I've overlooked any context, please feel free to provide clarification :)

Thank you all for your great work!

**Comment by younesbelkada** [comment]
Hi @mjbooo 
Thanks for the issue!
the case `if self.ref_model is None:` is only applied when one trains PEFT adapters. In that case, we call the `disable_adapters()` context manager to retrieve the logits of the model without the lora adapters that will act as the reference model. See: https://huggingface.co/blog/trl-peft for more details
